edb/edgeql/parser/grammar/commondl.py

```diff
@@ -543,3 +561,111 @@ def reduce_ExtensionVersion(self, *kids):
 
     def reduce_empty(self, *kids):
         self.val = None
+
+class IndexArg(Nonterm):
+    def reduce_kwarg_definition(self, *kids):
+        r"""%reduce FuncDeclArgName COLON \
+                OptTypeQualifier FullTypeExpr OptDefault \
+        """
+        self.val = qlast.FuncParam(
+            kind=qltypes.ParameterKind.PositionalParam,
+            name=kids[0].val,
+            typemod=kids[2].val,
+            type=kids[3].val,
+            default=kids[4].val
+        )
+
+    def reduce_AnyIdentifier_ASSIGN_Expr(self, *kids):
+        self.val = (
+            kids[0].val,
+            kids[0].context,
+            kids[2].val,
+        )
+
+    def reduce_FuncDeclArgName_OptDefault(self, *kids):
+        raise EdgeQLSyntaxError(
+            f'missing type declaration for the `{kids[0].val}` parameter',
+            context=kids[0].context)
+
+
+class IndexArgList(ListNonterm, element=IndexArg, separator=tokens.T_COMMA):
+    pass
+
+
+class OptIndexArgList(Nonterm):
+    def reduce_IndexArgList_COMMA(self, *kids):
+        self.val = kids[0].val
+
+    def reduce_IndexArgList(self, *kids):
+        self.val = kids[0].val
+
+    def reduce_empty(self, *kids):
+        self.val = []
+
+
+class IndexExtArgList(Nonterm):
+
+    def reduce_LPAREN_OptIndexArgList_RPAREN(self, *kids):
+        self.val = kids[1].val
+
+
+class OptIndexExtArgList(Nonterm):
+
+    def reduce_IndexExtArgList(self, *kids):
+        self.val = kids[0].val
+
+    def reduce_empty(self):
+        self.val = []
+
+
+class ProcessIndexArgsMixin:
+    def _process_arguments(self, arguments):
```

Comment:
This could just be a function, I think, instead of a mixin?

Reply:
Oh but there's an existing one for functions?

Reply:
At the moment I'm just following the same pattern as for functions.